[Enhancement] optimize lookup for dictionary

srlch opened the pull request:
[Enhancement] optimize lookup for dictionary
    
This pr introduce some optimization for looking up dictionary:
1. reserve the memory for the dest column which using for saving the result
   to avoid the memory reallocation.
2. Using _append_value instead of creating Datum for appending.
3. If the value type is Slice, save the slice first and append when to loop
   is finished. Avoid polluting the cpu cache for the prefetched content.

In our test case, for 1BE(16cpu, 128GB) + 1FE, here is the expression computation time:
key: int -> value: (int) : 572ms(before optimization) vs. 268ms(after optimization)
key: int -> value: (int, int) : 732ms(before optimization) vs. 395ms(after optimization)
key: int -> value: (int, int, varchar(length 128)) : 1s211ms(before optimization) vs. 744ms(after optimization)


Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

trueeyu commented on be/src/storage/dictionary_cache_manager.h:
```diff
@@ -155,7 +167,15 @@ class DictionaryCacheImpl final : public DictionaryCache {
+            DCHECK(value_encode_flags.size() == size);
+            // avoid memory reallocation when looking up hash table
+            if constexpr (std::is_same_v<ValueCppType, Slice>) {
+                down_cast<BinaryColumn*>(dest)->reserve(size, _max_slice_size * size);
```
Can't I get an exact value here?

srlch replied:
updated

trueeyu commented on be/src/storage/dictionary_cache_manager.h:
```diff
@@ -190,9 +223,42 @@ class DictionaryCacheImpl final : public DictionaryCache {
                         if constexpr (std::is_same_v<ValueCppType, Slice>) {
-                            value_encode_flags[i] = *(reinterpret_cast<uint8_t*>(iter->second.data) - 1);
+                            slices[beg_index + j].data = iter->second.data;
+                            slices[beg_index + j].size = iter->second.size;
```
Fetch the real reserve size?

srlch replied:
updated

srlch committed:
[Enhancement] optimize lookup for dictionary

This pr introduce some optimization for looking up dictionary:
1. reserve the memory for the dest column which using for saving the result
   to avoid the memory reallocation.
2. Using _append_value instead of creating Datum for appending.
3. If the value type is Slice, save the slice first and append when to loop
   is finished. Avoid polluting the cpu cache for the prefetched content.

In our test case, for 1BE(16cpu, 128GB) + 1FE, here is the expression computation time:
key: int -> value: (int) : 572ms(before optimization) vs. 268ms(after optimization)
key: int -> value: (int, int) : 732ms(before optimization) vs. 395ms(after optimization)
key: int -> value: (int, int, varchar(length 128)) : 1s211ms(before optimization) vs. 744ms(after optimization)

Signed-off-by: srlch <linzichao@starrocks.com>

trueeyu commented on be/src/storage/dictionary_cache_manager.h:
```diff
@@ -234,15 +283,29 @@ class DictionaryCacheImpl final : public DictionaryCache {
-        case DictionaryCacheEncoderType::PK_ENCODE: {
-            return std::make_shared<Datum>(v);
+    // Avoid creating Datum
+    inline void _append_value(Column* dest, const ValueCppType& v) {
```
Why not write like this?     

```
template <LogicalType type>
    inline void _append_value(Column* dest, const RunTimeTypeTraits<type>& v) {
        down_cast<RunTimeTypeTraits<type>>(dest)->append(v);
    }
```

srlch replied:
Updated

srlch committed:
Fix comment

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix test

Signed-off-by: srlch <linzichao@starrocks.com>
